lib/core/cmap/message_stream.js

```diff
@@ -0,0 +1,177 @@
+'use strict';
+
+const Duplex = require('stream').Duplex;
+const BufferList = require('bl');
+const MongoParseError = require('../error').MongoParseError;
+const decompress = require('../wireprotocol/compression').decompress;
+const Response = require('../connection/commands').Response;
+const BinMsg = require('../connection/msg').BinMsg;
+const MongoError = require('../error').MongoError;
+const OP_COMPRESSED = require('../wireprotocol/shared').opcodes.OP_COMPRESSED;
+const OP_MSG = require('../wireprotocol/shared').opcodes.OP_MSG;
+const MESSAGE_HEADER_SIZE = require('../wireprotocol/shared').MESSAGE_HEADER_SIZE;
+const COMPRESSION_DETAILS_SIZE = require('../wireprotocol/shared').COMPRESSION_DETAILS_SIZE;
+const opcodes = require('../wireprotocol/shared').opcodes;
+const compress = require('../wireprotocol/compression').compress;
+const compressorIDs = require('../wireprotocol/compression').compressorIDs;
+const uncompressibleCommands = require('../wireprotocol/compression').uncompressibleCommands;
+const Msg = require('../connection/msg').Msg;
+
+const kDefaultMaxBsonMessageSize = 1024 * 1024 * 16 * 4;
+const kBuffer = Symbol('buffer');
+
+/**
+ * A duplex stream that is capable of reading and writing raw wire protocol messages, with
+ * support for optional compression
+ */
+class MessageStream extends Duplex {
+  constructor(options) {
+    options = options || {};
+    super(options);
+
+    this.bson = options.bson;
+    this.maxBsonMessageSize = options.maxBsonMessageSize || kDefaultMaxBsonMessageSize;
+
+    this[kBuffer] = new BufferList();
+  }
+
+  _write(chunk, _, callback) {
+    this[kBuffer].append(chunk);
```

Comment:
if the reference stays the same, maybe pull it off into a const for readability:

```js
const buffer = this[kBuffer];

buffer.append(chunk);
...
```